src/libraries/System.Text.Json/tests/TrimmingTests/Helper.cs

```diff
@@ -0,0 +1,138 @@
+        {
+            JsonValueKind valueKind = expected.ValueKind;
+            if (valueKind != actual.ValueKind)
+            {
+                return false;
+            }
+
+            switch (valueKind)
+            {
+                case JsonValueKind.Object:
+                    var propertyNames = new HashSet<string>();
+
+                    foreach (JsonProperty property in expected.EnumerateObject())
+                    {
+                        propertyNames.Add(property.Name);
+                    }
+
+                    foreach (JsonProperty property in actual.EnumerateObject())
+                    {
+                        propertyNames.Add(property.Name);
+                    }
+
+                    foreach (string name in propertyNames)
+                    {
+                        if (!JsonEqual(expected.GetProperty(name), actual.GetProperty(name)))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                case JsonValueKind.Array:
+                    JsonElement.ArrayEnumerator expectedEnumerator = actual.EnumerateArray();
+                    JsonElement.ArrayEnumerator actualEnumerator = expected.EnumerateArray();
+
+                    while (expectedEnumerator.MoveNext())
+                    {
+                        if (!actualEnumerator.MoveNext())
+                        {
+                            return false;
+                        }
+
+                        if (!JsonEqual(expectedEnumerator.Current, actualEnumerator.Current))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return !actualEnumerator.MoveNext();
+                case JsonValueKind.String:
+                    return expected.GetString() == actual.GetString();
+                case JsonValueKind.Number:
+                case JsonValueKind.True:
+                case JsonValueKind.False:
+                case JsonValueKind.Null:
+                    return expected.GetRawText() == actual.GetRawText();
+                default:
+                    throw new NotSupportedException($"Unexpected JsonValueKind: JsonValueKind.{valueKind}.");
+            }
+        }
+
+        /// <summary>
+        /// Verifies the result of deserialization by serializing and comparing the output
+        /// with the expected payload. With this call pattern, the serialize should preserve
+        /// properties on typeof(object), but that would not be helpful to the calling test.
+        /// </summary>
+        public static bool VerifyWithSerialize(object obj, string expected)
+        {
+            return JsonEqual(expected, JsonSerializer.Serialize(obj));
+        }
+
+        public static bool AssertCollectionAndSerialize<T>(object obj, string json)
+        {
+            return obj is T && JsonSerializer.Serialize(obj) == json;
```

Comment:
This may be a bit of a stretch, but are we concerned that when we are testing `Deserialize` we are also calling `Serialize`? Could there ever be a bug where if the app ONLY called `Deserialize`, something was broken? But we are always doing both in our tests.

Reply:
The intention of these collection tests is to exercise the warm-up code paths of the serializer which is the same for both serialization and deserialization, so we can test either side. Here, we are checking that the reflection-based dependencies like the collection converters, `CreateRange` methods for immutable collections, `Push`/`Enqueue` for `Stack`/`Queue` etc., are rooted when needed.

The tests for each serializer method overload are found in the `TrimmingTest\SerializerEntryPoint` directory. For each of these, we test a POCO and collection. Issues with serialization or deserialization specifically would be pointed out by these tests.